File: camera_depth.py
import ArducamDepthCamera as ac
import numpy as np
from scipy.spatial.transform import Rotation as R
import open3d as o3d

MAX_DISTANCE = 4000

def open_depth_camera():
    cam = ac.ArducamCamera()
    
    if cam.open(ac.Connection.CSI, 8) != 0:
        raise RuntimeError("Failed to open Arducam depth camera at index 8")
    
    if cam.start(ac.FrameType.DEPTH) != 0:
        cam.close()
        raise RuntimeError("Failed to start Arducam depth camera")

    cam.setControl(ac.Control.RANGE, MAX_DISTANCE)
    return cam

def capture_depth_frame(cam, retries=10):
    for attempt in range(retries):
        frame = cam.requestFrame(2000)
        if frame is not None:
            return frame.depth_data, frame.confidence_data
        print(f"⚠️ Retry {attempt + 1}: Failed to capture depth frame")
    raise RuntimeError("Failed to capture frame from depth camera after retries")

def get_intrinsics(cam):
    fx = cam.getControl(ac.Control.INTRINSIC_FX) / 100.0
    fy = cam.getControl(ac.Control.INTRINSIC_FY) / 100.0
    cx = cam.getControl(ac.Control.INTRINSIC_CX) / 100.0
    cy = cam.getControl(ac.Control.INTRINSIC_CY) / 100.0

    K = np.array([[fx, 0, cx],
                  [0, fy, cy],
                  [0,  0,  1]])
    return K

def depth_to_pointcloud(depth_map, K, scale=1000.0):
    h, w = depth_map.shape
    fx, fy = K[0, 0], K[1, 1]
    cx, cy = K[0, 2], K[1, 2]

    i, j = np.meshgrid(np.arange(w), np.arange(h))
    z = depth_map / scale
    x = (i - cx) * z / fx
    y = (j - cy) * z / fy

    points = np.stack((x, y, z), axis=-1).reshape(-1, 3)
    valid = np.isfinite(points[:, 2]) & (points[:, 2] > 0)
    return points[valid]

def save_pointcloud_ply(points, filename="pointcloud.ply"):
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points)
    o3d.io.write_point_cloud(filename, pcd)
    print(f"✅ Saved point cloud to {filename}")

if __name__ == "__main__":
    cam = open_depth_camera()
    print("Camera opened and started successfully.")
    
    intrinsics = get_intrinsics(cam)
    print("Intrinsics:\n", intrinsics)

    depth, confidence = capture_depth_frame(cam)
    print("Captured depth shape:", depth.shape)
    print("Captured confidence shape:", confidence.shape)

    # Generate point cloud
    pointcloud = depth_to_pointcloud(depth, intrinsics)
    print("Generated point cloud with", pointcloud.shape[0], "points.")

    # Save to file
    save_pointcloud_ply(pointcloud)

    cam.stop()
    cam.close()

File: cpp_optimized/src/camera_depth.cpp
// camera_depth.cpp
#include "camera_depth.hpp"
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <iostream>

#define MAX_DISTANCE 4000

DepthCamera::DepthCamera() : opened_(false) {}

bool DepthCamera::open() {
    if (cam_.open(Arducam::Connection::CSI, 8) != 0) {
        std::cerr << "❌ Failed to open Arducam Depth Camera.\n";
        return false;
    }
    cam_.setControl(Arducam::Control::RANGE, MAX_DISTANCE);
    opened_ = true;
    return true;
}

bool DepthCamera::start() {
    return opened_ && cam_.start(Arducam::FrameType::DEPTH_FRAME) == 0;
}

bool DepthCamera::stop() {
    return opened_ && cam_.stop() == 0;
}

bool DepthCamera::close() {
    return opened_ && cam_.close() == 0;
}

bool DepthCamera::capture(cv::Mat& depth_map) {
    if (!opened_) return false;
    Arducam::ArducamFrameBuffer* frame = cam_.requestFrame(2000);
    if (!frame) return false;

    Arducam::FrameFormat format;
    frame->getFormat(Arducam::FrameType::DEPTH_FRAME, format);
    int width = format.width;
    int height = format.height;

    float* data = (float*)frame->getData(Arducam::FrameType::DEPTH_FRAME);
    depth_map = cv::Mat(height, width, CV_32F, data).clone();

    cam_.releaseFrame(frame);
    return true;
}

cv::Mat DepthCamera::getIntrinsicsMatrix() {
    int fx_raw, fy_raw, cx_raw, cy_raw;
    cam_.getControl(Arducam::Control::INTRINSIC_FX, &fx_raw);
    cam_.getControl(Arducam::Control::INTRINSIC_FY, &fy_raw);
    cam_.getControl(Arducam::Control::INTRINSIC_CX, &cx_raw);
    cam_.getControl(Arducam::Control::INTRINSIC_CY, &cy_raw);

    float fx = fx_raw / 100.0f;
    float fy = fy_raw / 100.0f;
    float cx = cx_raw / 100.0f;
    float cy = cy_raw / 100.0f;

    return (cv::Mat_<float>(3, 3) << fx, 0, cx,
                                    0, fy, cy,
                                    0,  0,  1);
}
